src/main/scala/com/microsoft/hyperspace/index/sources/delta/DeltaLakeRelation.scala

```diff
@@ -144,4 +145,94 @@ class DeltaLakeRelation(spark: SparkSession, override val plan: LogicalRelation)
+      }
+    }
+  }
+
+  /**
+   * Returns IndexLogEntry of the closest index version for the given relation.
+   *
+   * curFiles is used to calculate the similarity with each index version data.
+   * Delta Lake source provider utilizes the history information of DELTA_VERSION_HISTORY_PROPERTY
+   * to find the closet version of the index.
+   *
+   * @param index Candidate index to be applied.
+   * @return IndexLogEntry of the closest version among available index versions.
+   */
+  override def closestIndexVersion(index: IndexLogEntry): IndexLogEntry = {
+    // Seq of (index log version, delta lake table version)
+    val versions = deltaLakeVersionHistory(index)
+
+    lazy val indexManager = Hyperspace.getContext(spark).indexCollectionManager
+    def getIndexLogEntry(logVersion: Int): IndexLogEntry = {
+      indexManager
+        .getIndexLogEntry(index.name, logVersion)
+        .get
+        .asInstanceOf[IndexLogEntry]
+    }
+    // TODO: Currently assume all versions of index data exist.
+    //  Need to check and remove candidate indexes.
+    //  See https://github.com/microsoft/hyperspace/issues/387
+
+    plan.relation match {
+      case HadoopFsRelation(location: TahoeLogFileIndex, _, _, _, _, _) =>
+        // Find the largest index version whose delta table version is equal or less than
+        // the given relation.
+        val equalOrLessThanLastIndex = versions.lastIndexWhere(location.tableVersion >= _._2)
+        if (equalOrLessThanLastIndex == versions.size - 1) {
+          // The given table version is equal or larger than the latest index's.
+          // Use the latest version.
+          index
+        } else if (equalOrLessThanLastIndex == -1) {
+          // The given table version is smaller than the version at index creation.
+          // Use the initial version.
+          getIndexLogEntry(versions.head._1)
+        } else if (versions(equalOrLessThanLastIndex)._2 == location.tableVersion) {
+          // There is the index version that built for the given table version.
+          // Use the exact version.
+          getIndexLogEntry(versions(equalOrLessThanLastIndex)._1)
+        } else {
+          // Now, there are 2 candidate versions for the given table version:
+          // prevPair(indexVer1, tablePrevVersion), nextPair(indexVer2, tableNextVersion)
+          // which is (tablePrevVersion < TimeTravel table version < tableNextVersion).
+          val prevPair = versions(equalOrLessThanLastIndex)
+          val nextPair = versions(equalOrLessThanLastIndex + 1)
+
+          val prevLog = getIndexLogEntry(prevPair._1)
+          val nextLog = getIndexLogEntry(nextPair._1)
+
+          def getAppendedAndDeletedBytes(logEntry: IndexLogEntry): (Long, Long) = {
+            val commonBytes =
+              allFileInfos.filter(logEntry.sourceFileInfoSet.contains).map(_.size).sum
+            (allFileSizeInBytes - commonBytes, logEntry.sourceFilesSizeInBytes - commonBytes)
```

Comment:
looking at the usage, you can just return the sum of extra bytes? Do you need to separate these?